### PR TITLE
xrpl.js 2.0 Documentation Updates

### PR DESCRIPTION
**Pull request by mDuo13**
- Updates code samples from ripple-lib 1.x to xrpl.js 2.0, taking advantage of new features (like reliable transaction submission) where possible
- Updates interactive tutorials to use xrpl.js 2.0
- Updates dev tools to use xrpl.js 2.0
- Replaces the name of the library in various mentions, and updates links to the reference docs
- Rewrites the "Get Started with JavaScript" example
- Adds a migration guide for ripple-lib 1.x to xrpl.js 2.0. (WIP)

Fixes:
- #1186 
- #1187 
- #1188
- #1125

### CODE REVIEW

**Comment by wojake** [comment]
Mind adding this https://github.com/XRPLF/xrpl-dev-portal/issues/1189?

File: content/_snippets/checkcash-prereqs.md [hunk 1]
```diff
@@ -5,6 +5,5 @@ The prerequisites for cashing a check are the same whether you are cashing it fo
 - The **address** and **secret key** of the Check's stated recipient. The address must match the `Destination` address in the Check object.
 - If the Check is for an issued currency, you (the recipient) must have a trust line to the issuer. Your limit on that trust line must be enough higher than the balance to add the amount you would receive.
```
Comment (amarantha-k) [inline content/_snippets/checkcash-prereqs.md]: Unless "enough higher" implies a quantifiable value, I'd suggest rewording "enough higher than the balance..." to "greater than the balance...".

File: content/concepts/issued-currencies/rippling.md [hunk 2]
```diff
@@ -75,19 +75,14 @@ For more information, see [Default Ripple in 'Becoming an XRP Ledger Gateway'](b
 
 The No Ripple flag can only be enabled on a trust line if the address has a positive or zero balance on that trust line. This is so that the feature cannot be abused to default on the obligation the trust line balance represents. (Of course, you can still default by abandoning the address.)
 
-In the [`rippled` APIs](rippled-api.html), you can enable the No Ripple flag by sending a [TrustSet transaction][] with the `tfSetNoRipple` flag. You can disable No Ripple (enable rippling) with the `tfClearNoRipple` flag.
-
-In [RippleAPI](rippleapi-reference.html), you can enable the No Ripple flag by sending a [Trustline transaction](rippleapi-reference.html#preparetrustline) transaction with the `ripplingDisabled` field of the trust line set to `true`. <!-- SPELLING_IGNORE: trustline -->
+To enable the No Ripple flag, send a [TrustSet transaction][] with the `tfSetNoRipple` flag. You can disable No Ripple (enable rippling) with the `tfClearNoRipple` flag instead.
```
Comment (amarantha-k) [inline content/concepts/issued-currencies/rippling.md]: "disable No Ripple (enable ripping)" is a bit confusing on the first read
Suggestion: "disable the No Ripple flag (thereby enable rippling) with the `tfClearNoRipple` flag."

File: content/concepts/issued-currencies/transfer-fees.md [hunk 3]
```diff
@@ -47,24 +47,19 @@ In this scenario, Salazar (the sender) holds EUR issued by ACME, and wants to de
...
+**Note:** The [fix1201 amendment](amendments.html), introduced in `rippled` v0.80.0 and enabled on 2017-11-14, lowered the maximum transfer fee to 100% from an effective limit of approximately 329% (based on the maximum size of a 32-bit integer). The ledger may still contain accounts with a transfer fee setting higher than 100% (a `TransferRate` of `2000000000`). Any transfer fees already set continue to apply at their stated rate.
 
-A financial institution can submit an [AccountSet transaction][] from its [issuing address][] to change the `TransferRate` for all its issued currencies.
+## Client Library Support
 
-You can check an account's `TransferRate` with the [account_info method][]. If the `TransferRate` is omitted, then that indicates no fee.
+Some [client libraries](client-libraries.html) have convenience functions for getting and setting `TransferRate` functions.
 
+**JavaScript:** ***TODO based on https://github.com/XRPLF/xrpl.js/pull/1720***
```
Comment (amarantha-k) [inline content/concepts/issued-currencies/transfer-fees.md]: Should the TODO be an author note/comment?

File: content/concepts/payment-system-basics/transaction-basics/transaction-cost.md [hunk 4]
```diff
@@ -118,15 +118,14 @@ Before signing a transaction, we recommend [looking up the current load-based tr
...
+
+- If the network's transaction cost goes up between auto-filling and submitting the transaction, the transaction may not be confirmed.
+    - To prevent a transaction from getting stuck in a state of being neither definitively confirmed or rejected, be sure to provide a `LastLedgerSequence` parameter so it eventually expires. Alternatively, you can try to [cancel a stuck transaction](about-canceling-a-transaction.html) by reusing the same `Sequence` number. See [reliable transaction submission](reliable-transaction-submission.html) for best practices.
+- You have to be careful that the automatically provided value isn't too high. You don't want to burn a large fee to send a small transaction.
+    - If you are using `rippled`, you can also use the `fee_mult_max` and `fee_div_max` parameters of the [sign method][] to set a limit to the load scaling you are willing to sign.
+    - Some client libraries (like xrpl.js and xrpl-py) have configurable maximum `Fee` values, and raise an error instead of signing a transaction whose `Fee` value is higher than the maximum.
```
Comment (amarantha-k) [inline content/concepts/payment-system-basics/transaction-basics/transaction-cost.md]: Add a link to xrpl.js?

File: content/references/xrpljs2-migration-guide.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+Alternatively, you can use the `sign` method of a wallet to sign a transaction and then use `submitAndWait(tx_blob)` to submit it. This can be useful for building [reliable transaction submission](reliable-transaction-submission.html) that can recover from power outages and other disasters. (The library does not handle disaster recovery on its own.)
+
+
+
+## Quick Reference of Equivalents
+
+In ripple-lib 1.x all methods and properties were on instances of the `RippleAPI` class. In xrpl.js 2.x, some methods are static methods of the library and some methods belong to specific classes.
+
+| RippleAPI instance method / property | xrpl.js method / property| Notes |
```
Comment (amarantha-k) [inline content/references/xrpljs2-migration-guide.md]: missing space: "xrpl.js method/property | Notes"

File: content/tutorials/get-started/get-started-using-javascript.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+showcase_icon: assets/img/logos/javascript.svg
+---
+# Get Started Using JavaScript
+
+This tutorial guides you through the basics of building an XRP Ledger-connected application in JavaScript or TypeScript using the [`xrpl.js`](https://github.com/XRPLF/xrpl.js/) client library in either Node.js or web browsers.
+
+The scripts and config files used in this guide are [available in this website's GitHub Repository]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/get-started/js/).
+
+
+## Learning goals
```
Comment (amarantha-k) [inline content/tutorials/get-started/get-started-using-javascript.md]: Learning Goals, as per the style guide?

File: content/tutorials/get-started/get-started-using-javascript.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+## Install with npm
+
+Start a new project by creating an empty folder, then change into that folder and use [NPM](https://www.npmjs.com/) to install the latest version of xrpl.js:
+
+```sh
+npm install xrpl
+```
+
+
+## Start building
```
Comment (amarantha-k) [inline content/tutorials/get-started/get-started-using-javascript.md]: "Building"

File: content/tutorials/get-started/get-started-using-javascript.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+### 2. Connect to the XRP Ledger
+
+To make queries and submit transactions, you need to establish a connection to the XRP Ledger. To do this with `xrpl.js`, you create an instance of the `Client` class and use the `connect()` method.
+
+**Tip:** Many network functions in `xrpl.js` use [Promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) to return values asynchronously. The code samples here use the [`async/await` pattern](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await) to wait for the actual result of the Promises.
+
+{{ include_code("_code-samples/get-started/js/base.js", language="js") }}
+
+#### Connect to the production XRP Ledger
```
Comment (amarantha-k) [inline content/tutorials/get-started/get-started-using-javascript.md]: "Production"

File: content/tutorials/get-started/get-started-using-javascript.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+```
+
+
+## Start building
+
+When you're working with the XRP Ledger, there are a few things you'll need to manage, whether you're adding XRP into your [wallet](wallets.html), integrating with the [decentralized exchange](decentralized-exchange.html), or [issuing tokens](issued-currencies.html). This tutorial walks you through basic patterns common to getting started with all of these use cases and provides sample code for implementing them.
+
+Here are some steps you use in many XRP Ledger projects:
+
+1. [Import the library](#1-import-the-library)
```
Comment (amarantha-k) [inline content/tutorials/get-started/get-started-using-javascript.md]: Missing period - "1. Import the Library."

File: content/tutorials/get-started/get-started-using-javascript.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+{{ include_code("_code-samples/get-started/js/base.js", language="js") }}
+
+#### Connect to the production XRP Ledger
+
+The sample code in the previous section shows you how to connect to the Testnet, which is one of the available [parallel networks](parallel-networks.html). When you're ready to integrate with the production XRP Ledger, you'll need to connect to the Mainnet. You can do that in two ways:
+
+* By [installing the core server](install-rippled.html) (`rippled`) and running a node yourself. The core server connects to the Mainnet by default, but you can [change the configuration to use Testnet or Devnet](connect-your-rippled-to-the-xrp-test-net.html). [There are good reasons to run your own core server](the-rippled-server.html#reasons-to-run-your-own-server). If you run your own server, you can connect to it like so:
+
+        const MY_SERVER = "ws://localhost:6006/"
```
Comment (amarantha-k) [inline content/tutorials/get-started/get-started-using-javascript.md:102]: Enclose this in a code block?

File: content/tutorials/get-started/get-started-using-javascript.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+        const client = new xrpl.Client(MY_SERVER)
+        await client.connect()
+
+    See the example [core server config file](https://github.com/ripple/rippled/blob/c0a0b79d2d483b318ce1d82e526bd53df83a4a2c/cfg/rippled-example.cfg#L1562) for more information about default values.
+
+* By using one of the available [public servers][]:
+
+        const PUBLIC_SERVER = "wss://xrplcluster.com/"
```
Comment (amarantha-k) [inline content/tutorials/get-started/get-started-using-javascript.md:110]: Enclose this in a code block?

File: content/tutorials/production-readiness/reliable-transaction-submission.md [hunk 12]
```diff
@@ -36,19 +36,19 @@ If a power or network outage occurs, applications face more challenges finding t
 
 ### Transaction Timeline
 
-The XRP Ledger provides several APIs for submitting transactions, including [`rippled` API](rippled-api.html), and [RippleAPI](rippleapi-reference.html).  Regardless of the API used, the transaction is applied to the ledger as follows.
+When you submit a transaction to the XRP Ledger, regardless of whether used [HTTP API](rippled-api.html), a [client library](client-libraries.html), or some other app, process of applying the transaction to the ledger is the same. That process goes like this:
```
Comment (amarantha-k) [inline content/tutorials/production-readiness/reliable-transaction-submission.md]: regardless of whether you used...or some other app, the process of applying....

File: content/tutorials/use-tokens/enable-no-freeze.md [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+parent: use-tokens.html
+blurb: Permanently give up your account's ability to freeze tokens it issues.
+filters:
+  - include_code
+labels:
+  - Tokens
+---
+# Enable No Freeze
+
+If you [issue tokens](issued-currencies.html) in the XRP Ledger, can enable the [No Freeze setting](freezes.html#no-freeze) to permanently limit your own ability to use the token freezing features of the XRP Ledger. (As a reminder, this only applies to issued tokens, not XRP.) This tutorial shows how to enable the No Freeze setting on your issuing account.
```
Comment (amarantha-k) [inline content/tutorials/use-tokens/enable-no-freeze.md:12]: Suggestion:
When you [issue tokens]...., you can enable the [No Freeze setting]...

File: content/tutorials/use-tokens/freeze-a-trust-line.md [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+    ]
+  }
+}
+```
+
+<!-- MULTICODE_BLOCK_END -->
+
+### {{n.next()}}. (Optional) Send TrustSet Transaction to End the Freeze
+
+If you decide that the trust line no longer needs to be frozen (for example, you investigated and decided that the suspicious activity was benign), you can end the individual freeze in much the same way that you froze the trust line in the first place. To end an individual freeze, send a [TrustSet transaction][] with the [`tfClearFreeze` flag enabled](trustset.html#trustset-flags). The other fields of the transaction should be the same as when you froze the trust line:
```
Comment (amarantha-k) [inline content/tutorials/use-tokens/freeze-a-trust-line.md]: ...,you can end the individual freeze in pretty much the same way that you froze...
or
...,you can end the individual freeze in the same way that you froze...

**Review by amarantha-k — COMMENTED** [review]
Submitted comments from a peer/editorial review of the content.

File: content/_snippets/checkcash-prereqs.md (hunk 1 above)
Comment (mDuo13) [reply content/_snippets/checkcash-prereqs.md]: You're right that the wording here is twisted; but saying "greater than the balance" is not quite right either because what actually matters is _the difference between the limit and the current balance_. For example, if your balance is 100 and your limit is 110, you don't have room to receive more than 10.

I'll reword to this:

> Your limit on that trust line must be high enough to hold your previous balance plus the amount you would receive.

File: content/concepts/issued-currencies/transfer-fees.md (hunk 3 above)
Comment (mDuo13) [reply content/concepts/issued-currencies/transfer-fees.md]: I've left TODOs in bold-italic in places where I expect/need to address them before publishing.

File: content/tutorials/get-started/get-started-using-javascript.md (hunk 10 above)
Comment (mDuo13) [reply content/tutorials/get-started/get-started-using-javascript.md:102]: Since this is part of a bulleted list, code fences are not (reliably) an option, but indenting the appropriate number of spaces works to achieve the same effect. In this case, it's indented 8 spaces: 4 spaces to continue the previous bullet point + an addition 4 spaces to mark it as a code block.

You can see it renders correctly [in the preview build](https://xrplf.github.io/xrpl-dev-portal/pr-preview/xrpljs2.0/get-started-using-javascript.html#connect-to-the-production-xrp-ledger):



See also: [lists and code blocks](https://dactyl.link/lists-and-codeblocks.html) [(raw markdown)](https://raw.githubusercontent.com/ripple/dactyl/master/examples/content/lists-and-codeblocks.md)

File: content/tutorials/use-tokens/enable-no-freeze.md (hunk 13 above)
Comment (mDuo13) [reply content/tutorials/use-tokens/enable-no-freeze.md:12]: I think "When you..." implies it has to happen at the same time as issuing the tokens. It can actually happen before or after. So I'd like to leave this one as-is.

File: content/tutorials/use-tokens/freeze-a-trust-line.md (hunk 14 above)
Comment (mDuo13) [reply content/tutorials/use-tokens/freeze-a-trust-line.md]: How about "almost the same way"? Since it's not _exactly_ the same (the flag value is different) and "pretty much" feels too informal to me.

File: content/references/xrpljs2-migration-guide.md [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+```
+
+
+## Validated Results
+
+By default, most methods in ripple-lib 1.x only returned results that were validated by the [consensus process](consensus.html) and therefore final. The xrpl.js equivalents of many methods use the [`Client.request()` method](https://js.xrpl.org/classes/Client.html#request) to call the WebSocket API, where the XRP Ledger server's default settings often use the current (pending) ledger to serve data which is not final.
+
+Sometimes you want to use the current open ledger because it has the pending results of many transactions that are likely to succeed, such as when looking up the state of the [decentralized exchange](decentralized-exchange.html). In other cases, you want to use a validated ledger, which only incorporates the results of transactions that are finalized.
+
+When making API requests with xrpl.js 2.0 using the , you should explicitly [specify what ledger to use](basic-data-types.html#specifying-ledgers). For example, to look up trust lines using the latest _validated ledger_:
```
Comment (JST5000) [inline content/references/xrpljs2-migration-guide.md]: Missing something before the comma: "When making API requests with xrpl.js 2.0 using the _________, you ..."

File: content/references/xrpljs2-migration-guide.md [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+const prepared = await client.autofill({
+  "TransactionType": "Payment",
+  "Account": sender,
+  "Amount": xrpl.xrpToDrops("50.2"),
+  "Destination": "rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe",
+  "LastLedgerSequence": vli+75 // gives ~5min, rather than the default ~1min
+})
+```
+
+Like the old prepare methods, `Client.autofill()` provides a reasonable `LastLedgerSequence` value by default. To prepare a transaction _without_ a `LastLedgerSequence` field, pass it with the value `null`:
```
Comment (JST5000) [inline content/references/xrpljs2-migration-guide.md:128]: If you don't set `LastLedgerSequence` at all, autofill will still fill it out. (Just tested it locally) - To clear the `LastLedgerSequence` you can overwrite it after autofilling.

Comment (JST5000) [reply content/references/xrpljs2-migration-guide.md:128]: Oops, I misread the intent here! (Even though you added emphasis on the _without_)

File: content/references/xrpljs2-migration-guide.md [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+  "streams": ["ledger"]
+})
+```
+
+
+## Reference of Equivalents
+
+In ripple-lib 1.x all methods and properties were on instances of the `RippleAPI` class. In xrpl.js 2.x, some methods are static methods of the library and some methods belong to specific classes. In the following table, the notation `Client.method()` means that `method()` belongs to instances of the `Client` class.
+
+| RippleAPI instance method / property | xrpl.js method / property | Notes |
```
Comment (JST5000) [inline content/references/xrpljs2-migration-guide.md:228]: On my Chrome browser I can't see the 3rd column (I didn't even realize it was there - Screenshot attached) 

Is there any way to limit the horizontal width to guarantee that at least the start of the third column is visible? 

image

Comment (JST5000) [reply content/references/xrpljs2-migration-guide.md:228]: Alternatively, instead of "(Removed)" - Could you add a comment there saying "Removed - See column on right for more details." (For more discoverability of the 3rd column

File: content/references/xrpljs2-migration-guide.md [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+  "account": "rf1BiGeXwwQoi8Z2ueFYTEXSwuJYfV2Jpn",
+  "ledger_index": "validated"
+})
+console.log(trustlines.result)
+```
+
+
+## Transaction Submission
+
+In xrpl.js there are specific helper functions for signing and submitting transactions and waiting for the XRP Ledger blockchain to confirm those transactions' final outcomes. Use the `submitAndWait(tx_json, wallet)` method to sign a prepared transaction with the given wallet and submit it, like in this example:
```
Comment (mvadari) [inline content/references/xrpljs2-migration-guide.md]: IMO you should explain `submit` vs `submitAndWait` here, instead of just in the code comments and below the code snippet. I almost missed the explanation under the code snippet.

File: content/references/xrpljs2-migration-guide.md [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+```js
+const tx_json = await client.autofill({
+  "TransactionType": "AccountSet",
+  "Account": wallet.address, // "wallet" is an instance of the Wallet class
+  "SetFlag": xrpl.AccountSetAsfFlags.asfRequireDest
+})
+try {
+  const submit_result = await client.submitAndWait(tx_json, wallet)
+  // submitAndWait() doesn't return until the transaction has a final result.
+  // Raises XrplError if the transaction doesn't get confirmed by the network.
```
Comment (mvadari) [inline content/references/xrpljs2-migration-guide.md:100]:
```suggestion
  // Raises XrplError if the transaction doesn't get confirmed by the network.
  // Does not handle disaster recovery.
```

File: content/references/xrpljs2-migration-guide.md (hunk 17 above)
Comment (mDuo13) [reply content/references/xrpljs2-migration-guide.md:228]: Unfortunately getting the tables to display well on a lot of devices is a really hard problem to solve. I changed it to "(Removed - see Notes column)" for now.

File: content/references/xrpljs2-migration-guide.md [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+| `getOrderbook()` | `Client.getOrderbook()` | |
+| `getSettings()` | (Removed - see Notes column) | Use [`Client.request()`](https://js.xrpl.org/classes/Client.html#request) to call the [account_info method][] instead. Use `xrpl.parseAccountRootFlags()` on the `Flags` field to get the boolean values of individual flag settings. **Warning:** Unlike `getSettings()`, `account_info` can return [results that are not validated and final](#validated-results). |
+| `getAccountInfo(address, options)` | (Removed - see Notes column) | Use [`Client.request()`](https://js.xrpl.org/classes/Client.html#request) to call the [account_info method][] instead. **Warning:** Unlike `getAccountInfo()`, `account_info` can return [results that are not validated and final](#validated-results). |
+| `getAccountObjects(address, options)` | (Removed - see Notes column) | Use [`Client.request()`](https://js.xrpl.org/classes/Client.html#request) to call the [account_info method][] instead. **Warning:** Unlike `getAccountObjects()`, `account_objects` can return [results that are not validated and final](#validated-results). |
```
Comment (mvadari) [inline content/references/xrpljs2-migration-guide.md]: typo:
```suggestion
| `getAccountObjects(address, options)` | (Removed - see Notes column) | Use [`Client.request()`](https://js.xrpl.org/classes/Client.html#request) to call the [account_objects method][] instead. **Warning:** Unlike `getAccountObjects()`, `account_objects` can return [results that are not validated and final](#validated-results). |
```

File: content/tutorials/get-started/get-started-using-javascript.md [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+
+
+## Requirements
+
+To follow this tutorial, you should have some familiarity with writing code in JavaScript and managing small JavaScript projects. In browsers, any modern web browser with JavaScript support should work fine. In Node.js, **version 14** is recommended. Node.js versions 12 and 16 are also regularly tested.
+
+
+## Install with npm
+
+Start a new project by creating an empty folder, then change into that folder and use [NPM](https://www.npmjs.com/) to install the latest version of xrpl.js:
```
Comment (mvadari) [inline content/tutorials/get-started/get-started-using-javascript.md]: "change into" is weird wording
```suggestion
Start a new project by creating an empty folder, then move into that folder and use [NPM](https://www.npmjs.com/) to install the latest version of xrpl.js:
```

File: content/_code-samples/send-xrp/js/send-xrp.js [hunk 22]
```diff
@@ -0,0 +1,52 @@
+// Dependencies for Node.js.
+// In browsers, use a <script> tag instead.
+if (typeof module !== "undefined") {
+  // gotta use var here because const/let are block-scoped to the if statement.
+  var xrpl = require('xrpl')
+}
+
+// Example credentials
+const wallet = xrpl.Wallet.fromSeed("sn3nxiW7v8KXzPzAqzyHXbSSKNuN9")
```
Comment (mvadari) [inline content/_code-samples/send-xrp/js/send-xrp.js:9]: To match with the Python code, should this have a `console.log(wallet.classicAddress)`?

File: content/_code-samples/require-destination-tags/package.json [hunk 23]
```diff
@@ -1,5 +1,6 @@
 {
   "dependencies": {
-    "ripple-lib": "^1.10.0"
+    "ripple-lib": "^1.10.0",
+    "xrpl": "^2.0.0-beta.5"
```
Comment (mvadari) [inline content/_code-samples/require-destination-tags/package.json]: This can be updated to `2.0.0` (and `ripple-lib` can be removed)
